README.md
# Steer Protocol Strategy Template - Assemblyscript
AssemblyScript based Strategy Template for the Steer Protocol.

Start building on [Steer Protocol](https://steer.finance) with our AssemblyScript template. Expand your smart contract capabilities with off-chain compute! More info can be found here: [Documentation](https://docs.steer.finance/steer-apps/writing-an-app)

This is a template example of how to make a liquidity provision strategy that acts similarly to a stop loss. The strategist picks a percent and recent candles are passed into the strategy. From the recent price, a position is made with the upper tick set to the current price, and the lower tick at the precent down from the current price, like a stop loss. This would likely not be a great strategy for most assets/situations, be smart if you want to use this strategy.

## Parameters

This strategy only accepts one parameter from the data connectors: any OHLC type object.

## Project Structure
Apps have three external functions that are used by the Steer system. Additional methods, classes, or varaibles can be used in conjunction with these required functions for any desired behavior. For more information please see the app interface. This design means, that as a developer, you only need to implement the methods which are required for the app to work.

Below are the significant files and folders which you will want to get familiar with:

```
├── assembly      // Source code for the app
├── build         // Output of the build process aka `yarn asbuild`
├── coverage      // Coverage report for testing
├── tests         // Test files with a built in test runner
├── asconfig.json // Assemblyscript config
├── index.js      // Javascript entrypoint for the app when running tests
├── package.json  // Dependencies for the app
```

## Project Setup
Once the repository has been cloned, you will need to install the project dependencies. This can be done via the following command:

```yarn install```

INFO
You will notice that there is a post-install script which will compile the ./assembly source folder and populate the ./build folder. This is done to make it easier to run the tests. We will cover this later.

After making chenges to the assemblyscript code, compile and build the wasm files with the command:

```yarn asbuild```

You will find a number of webassembly related files in the ./build folder.

The testing suite can be found the ./tests source folder, unit tests are written in index.test.ts and can be configured there.

To run the testing suite, run:

```yarn test```

For more in depth testing of the strategy and performance try using Steer's backtesting toolkit to simulate realistic conditions.


CHANGELOG.md
## [1.3.0](https://github.com/SteerProtocol/strategy-template-assemblyscript/compare/v1.2.0...v1.3.0) (2023-04-13)


### Features

* :zap: f64, rename ([400500f](https://github.com/SteerProtocol/strategy-template-assemblyscript/commit/400500f8efe2145e56402461057fcd04f8758202))

## [1.2.0](https://github.com/SteerProtocol/strategy-template-assemblyscript/compare/v1.1.1...v1.2.0) (2023-03-21)


### Features

* :art: app-loader version ([af03954](https://github.com/SteerProtocol/strategy-template-assemblyscript/commit/af03954a51621bdb9961c133a049b61f01bf63f7))
* :fire: As conversion, simplified example ([b823d6e](https://github.com/SteerProtocol/strategy-template-assemblyscript/commit/b823d6e903f20b13b49b8f934c945b8ad863544d))
* :zap: rehaul-completion ([dc9950d](https://github.com/SteerProtocol/strategy-template-assemblyscript/commit/dc9950d72af32d283c83a762e03f7b3fa6dd6bb5))


tests/utils.ts
export const config_payload = `{"percent":5,"poolFee":3000}`;

export const config = `{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Strategy Config",
  "type": "object",
  "expectedDataTypes": ["OHLC"],
  "properties": {
    "percent": {
      "type": "number",
      "description": "Percent for trailing stop order (must be greater than pool fee)",
      "default": 5.0
  },
    "poolFee": {
      "description": "Pool fee percent for desired Uniswapv3 pool",
      "enum" : [10000, 3000, 500, 100],
      "enumNames": ["1%", "0.3%", "0.05%", "0.01%"]
    }
  },
  "required": ["percent", "poolFee"]
}`;

export const empty = '{"data":[]}';

export const prices2 = [
  {
    open: 2.50463873386213e21,
    high: 2.5046430693437536e21,
    close: 2.5046430693437536e21,
    low: 2.50463873386213e21,
  },
  {
    open: 2.5046441722633493e21,
    high: 2.504653822829505e21,
    close: 2.504653822829505e21,
    low: 2.5046441722633493e21,
  },
  {
    open: 2.5046560286781625e21,
    high: 2.5046687506200826e21,
    close: 2.5046687506200826e21,
    low: 2.5046560286781625e21,
  },
  {
    open: 2.504675500042307e21,
    high: 2.504675500042307e21,
    close: 2.504675500042307e21,
    low: 2.504675500042307e21,
  },
  {
    open: 2.5046794323890526e21,
    high: 2.5046794323890526e21,
    close: 2.5046794323890526e21,
    low: 2.5046794323890526e21,
  },
  {
    open: 2.5047014906624753e21,
    high: 2.5047014906624753e21,
    close: 2.5047014906624753e21,
    low: 2.5047014906624753e21,
  },
  {
    open: 2.5047064487510734e21,
    high: 2.5047064487510734e21,
    close: 2.5047064487510734e21,
    low: 2.5047064487510734e21,
  },
  {
    open: 2.50470920324877e21,
    high: 2.5047126739199585e21,
    close: 2.5047126739199585e21,
    low: 2.50470920324877e21,
  },
  {
    open: 2.504713775721294e21,
    high: 2.504713775721294e21,
    close: 2.504713775721294e21,
    low: 2.504713775721294e21,
  },
  {
    open: 2.5047159793253357e21,
    high: 2.5047159793253357e21,
    close: 2.5047159793253357e21,
    low: 2.5047159793253357e21,
  },
  {
    open: 2.5047181829312254e21,
    high: 2.5047181829312254e21,
    close: 2.5047181829312254e21,
    low: 2.5047181829312254e21,
  },
  {
    open: 2.504718733832986e21,
    high: 2.504718733832986e21,
    close: 2.504718733832986e21,
    low: 2.504718733832986e21,
  },
  {
    open: 2.504721488343512e21,
    high: 2.504736417840634e21,
    close: 2.504736417840634e21,
    low: 2.504721488343512e21,
  },
  {
    open: 2.5047366657481213e21,
    high: 2.5047369136556406e21,
    close: 2.5047369136556406e21,
    low: 2.5047366657481213e21,
  },
  {
    open: 2.504737161563177e21,
    high: 2.504737161563177e21,
    close: 2.504737161563177e21,
    low: 2.504737161563177e21,
  },
  {
    open: 2.5047373764163986e21,
    high: 2.5047492341717213e21,
    close: 2.5047492341717213e21,
    low: 2.5047373764163986e21,
  },
  {
    open: 2.5047501877966684e21,
    high: 2.5047501877966684e21,
    close: 2.5047501877966684e21,
    low: 2.5047501877966684e21,
  },
  {
    open: 2.5047519507197807e21,
    high: 5.5047574598621116e21,
    close: 2.5047574598621116e21,
    low: 2.5047519507197807e21,
  },
  {
    open: 2.5047927499044306e21,
    high: 2.5047927499044306e21,
    close: 2.5047927499044306e21,
    low: 2.5047927499044306e21,
  },
];

export const prices = [
  {
    "timestamp": 1620248400000,
    "high": 3.0482211045024987e-12,
    "low": 3.0420331741444116e-12,
    "close": 3.0482211045024987e-12,
    "open": 3.0420331741444116e-12
  },
  {
    "timestamp": 1620249300000,
    "high": 3.052443706603611e-12,
    "low": 3.051083443544758e-12,
    "close": 3.052443706603611e-12,
    "open": 3.051083443544758e-12
  },
  {
    "timestamp": 1620250200000,
    "high": 3.058971994885798e-12,
    "low": 3.053301697638979e-12,
    "close": 3.058971994885798e-12,
    "open": 3.053301697638979e-12
  },
  {
    "timestamp": 1620251100000,
    "high": 3.0619150435330332e-12,
    "low": 3.0591052152995406e-12,
    "close": 3.0619150435330332e-12,
    "open": 3.0591052152995406e-12
  },
  {
    "timestamp": 1620252000000,
    "high": 3.062551448212578e-12,
    "low": 3.0624942591314255e-12,
    "close": 3.062551448212578e-12,
    "open": 3.0624942591314255e-12
  },
  {
    "timestamp": 1620252900000,
    "high": 3.0777357130785475e-12,
    "low": 3.0642196617996214e-12,
    "close": 3.064368586375334e-12,
    "open": 3.0642196617996214e-12
  },
  {
    "timestamp": 1620253800000,
    "high": 3.0569112254326672e-12,
    "low": 3.0569112254326672e-12,
    "close": 3.0569112254326672e-12,
    "open": 3.0569112254326672e-12
  },
  {
    "timestamp": 1620254700000,
    "high": 3.053931783734179e-12,
    "low": 3.0317150735664757e-12,
    "close": 3.0317150735664757e-12,
    "open": 3.053931783734179e-12
  },
  {
    "timestamp": 1620255600000,
    "high": 3.0372917675150144e-12,
    "low": 3.0372917675150144e-12,
    "close": 3.0372917675150144e-12,
    "open": 3.0372917675150144e-12
  },
  {
    "timestamp": 1620256500000,
    "high": 3.0386120996797003e-12,
    "low": 3.0386120996797003e-12,
    "close": 3.0386120996797003e-12,
    "open": 3.0386120996797003e-12
  },
  {
    "timestamp": 1620261000000,
    "high": 3.0404990351591258e-12,
    "low": 3.0404990351591258e-12,
    "close": 3.0404990351591258e-12,
    "open": 3.0404990351591258e-12
  },
  {
    "timestamp": 1620261900000,
    "high": 3.0547701056451405e-12,
    "low": 3.0429306594344523e-12,
    "close": 3.0547701056451405e-12,
    "open": 3.0429306594344523e-12
  },
  {
    "timestamp": 1620262800000,
    "high": 3.0737543374303936e-12,
    "low": 3.0737543374303936e-12,
    "close": 3.0737543374303936e-12,
    "open": 3.0737543374303936e-12
  },
  {
    "timestamp": 1620264600000,
    "high": 3.0770023498727197e-12,
    "low": 3.074909951263263e-12,
    "close": 3.0770023498727197e-12,
    "open": 3.074909951263263e-12
  },
  {
    "timestamp": 1620265500000,
    "high": 3.076334560147718e-12,
    "low": 3.076334560147718e-12,
    "close": 3.076334560147718e-12,
    "open": 3.076334560147718e-12
  }
]

tests/index.test.ts
import { config, config_payload, prices2 } from "./utils";
import fs from 'fs';
import { WasmModule, loadWasm } from "@steerprotocol/app-loader";

describe("WASM Module", () => {
  let myModule: WasmModule;
  
  beforeEach(async () => {
    myModule = await loadWasm(fs.readFileSync(__dirname + "/../build/debug.wasm"), {})
  });
  describe("Custom Strategy", () => {
    test("can render config", async () => {
      // Call the config function on the strategy bundle
      const result = myModule.config();
      // Pull the result from memory and parse the result
      const parsedResult = JSON.parse(result);
      // The result should match the given config
      expect(parsedResult).toStrictEqual(JSON.parse(config));
    });

    test("can run execute", async () => {
      // The actual strategy instantiation and execution
      myModule.initialize(config_payload);
      // Call the config function on the strategy bundle
      const result = myModule["execute(param_1: string)"](JSON.stringify(prices2));
      // Pull the result from memory and parse the result
      const parsedResult = JSON.parse(result);
      // The result should match the given config
      expect(JSON.stringify(parsedResult)).toStrictEqual(
        `{\"functionName\":\"tend(uint256,(int24[],int24[],uint16[]),bytes)\",\"typesArray\":[\"uint256\",\"tuple(int24[],int24[],uint16[])\",\"bytes\"],\"valuesArray\":[10000,[[492180],[492780],[100]],\"0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000ffffffffffffffffffffffffffffffffffffffff\"]}`
      );
    });
  });
});



assembly/CustomStrategy.ts
import { BandsResult, UniswapV3LiquidityStrategy } from "@steerprotocol/concentrated-liquidity-strategy";
import { Price } from "@steerprotocol/strategy-utils";

export class CustomStrategy extends UniswapV3LiquidityStrategy {
  private result: BandsResult = [];

  public percent: f32 = 0;

  constructor(binWidth: f32, percent: f32) {
    // Setup bin widths for liquidity buckets
    // because we are extending the UniswapV3LiquidityStrategy
    super(binWidth);
    this.percent = percent;
  }

  trailingStop(prices: Price[]): f32 {
    // Get the current price of the asset pair
    const currentPrice = prices[prices.length - 1];
  
    // Calculate the trailing stop price
    const trailingStopPrice = currentPrice.close - (currentPrice.close * (this.percent / 100));
  
    // Return the trailing stop price
    return trailingStopPrice;
  }
}

assembly/index.ts
import { getTickFromPrice, getTickSpacing, renderULMResult } from "@steerprotocol/concentrated-liquidity-strategy/assembly";
import { parseCandles, Position, console, trailingStop } from  "@steerprotocol/strategy-utils/assembly";
import { JSON } from "json-as/assembly";

let percent: f64 = 0;
let poolFee: i32 = 0;

@serializable
class Config {
  poolFee: f64 = 0;
  percent: i32 = 0;
}

export function initialize(config: string): void {
  // Parse the config object
  const configJson: Config = JSON.parse<Config>(config);

  // Handle null case
  if (
    configJson.percent == 0 ||
    configJson.poolFee == 0 
  ) {
    throw new Error("Invalid configuration");
  }

  // Assign values to memory
  percent = f64(configJson.percent);
  poolFee = i32(configJson.poolFee);
}

function closestDivisibleNumber(num: number, divisor: number, floor: boolean): number {
  if (floor) return Math.floor(num / divisor) * divisor;
  return Math.ceil(num / divisor) * divisor;
}

export function execute(_prices: string): string {
  // _prices will have the results of the dc, which is only candles here
  const prices = parseCandles(_prices);
  // If we have no candles, skip action
  if (prices.length == 0) {
    return `continue`;
  }

  const lowerLimit = trailingStop(percent, prices);
  const upperLimit = prices[prices.length - 1].close;
  
  const upperTick = closestDivisibleNumber(i32(Math.round(getTickFromPrice(f64(upperLimit)))), getTickSpacing(poolFee), false);
  const lowerTick = closestDivisibleNumber(i32(Math.round(getTickFromPrice(f64(lowerLimit)))), getTickSpacing(poolFee), true);
  
  // Calculate position

  
  const positions = [new Position(i32(lowerTick), i32(upperTick), 100)];

  // Format and return result
  return renderULMResult(positions, 10000);
}

export function config(): string {
  return `{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Strategy Config",
    "type": "object",
    "expectedDataTypes": ["OHLC"],
    "properties": {
      "percent": {
        "type": "number",
        "description": "Percent for trailing stop order (must be greater than pool fee)",
        "default": 5.0
    },
      "poolFee": {
        "description": "Pool fee percent for desired Uniswapv3 pool",
        "enum" : [10000, 3000, 500, 100],
        "enumNames": ["1%", "0.3%", "0.05%", "0.01%"]
      }
    },
    "required": ["percent", "poolFee"]
  }`;
}

export function version(): i32 {
  return 1;
}
